CMakeLists.txt

```diff
@@ -74,15 +74,47 @@ if(BUILD_COMPILER STREQUAL "gcc" OR BUILD_COMPILER STREQUAL "clang")
     endif()
 endif()
 
+#specify linker options
+include(CheckLinkerFlag)
+check_linker_flag(CXX "-fuse-ld=gold" COMPILER_RT_HAS_FUSE_LD_GOLD_FLAG)
+check_linker_flag(CXX "-fuse-ld=lld" COMPILER_RT_HAS_FUSE_LD_LLD_FLAG)
+
+set(USE_LD_GOLD On CACHE BOOL "Use GNU gold linker")
+set(USE_LD_LLD ON CACHE BOOL "Use LLD linker")
+
 if(BUILD_COMPILER STREQUAL "gcc")
     add_compile_definitions($<$<CONFIG:Debug>:_GLIBCXX_ASSERTIONS>) # Enable assertions in libstdc++.
-    add_link_options(-fuse-ld=gold) # Use gold linker, faster than ld.
+    if(USE_LD_LLD AND COMPILER_RT_HAS_FUSE_LD_LLD_FLAG)
+        message(STATUS "Using lld linker.")
+        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fuse-ld=lld")
+    elseif(USE_LD_GOLD AND COMPILER_RT_HAS_FUSE_LD_GOLD_FLAG)
+        message(STATUS "Using gold linker.")
+        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fuse-ld=gold")
+    else()
+        message(STATUS "Fallback to default linker.")
+    endif()
+    # -Og is a better choice than -O0 for producing debuggable code because some compiler passes that collect debug information are disabled at -O0. 
+    add_compile_options(
+        "$<$<CONFIG:DEBUG>:-Og>"
+    )    
 elseif(BUILD_COMPILER STREQUAL "clang")
     add_compile_options(-Werror=unused-comparison) # Comparison result unused, not available in gcc.
     add_compile_options(-Werror=inconsistent-missing-override) # Clang-specific, goes with -Werror=suggest-override.
     add_compile_options(-Werror=invalid-noreturn) # function declared 'noreturn' should not return.
     add_compile_options(-Werror=uninitialized) # variable 'x' is used uninitialized.
     enable_libcxx_assertions(FALSE)
+    if(USE_LD_LLD AND COMPILER_RT_HAS_FUSE_LD_LLD_FLAG)
+        message(STATUS "Using lld linker.")
+        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fuse-ld=lld")
+    elseif(USE_LD_GOLD AND COMPILER_RT_HAS_FUSE_LD_GOLD_FLAG)
+        message(STATUS "Using gold linker.")
+        set(CMAKE_EXE_LINKER_FLAGS "${CMAKE_EXE_LINKER_FLAGS} -fuse-ld=gold")
+    else()
+        message(STATUS "Fallback to default linker.")
+    endif()
```

Comment:
Identical code blocks should go to inside the `if` above – `if(BUILD_COMPILER STREQUAL "gcc" OR BUILD_COMPILER STREQUAL "clang")`.